pkg/eosclient/eosgrpc/eos_http/eoshttp.go

```diff
@@ -0,0 +1,503 @@
+	// hence we don't need to lock it when used
+	defer httpTransportMtx.Unlock()
+
+	if httpTransport == nil {
+
+		// TODO: the error reporting of http.transport is insufficient
+		// must check manually at least the existence of the certfiles
+		// The point is that also the error reporting of the context that calls this function
+		// is weak
+		httpTransport = &http.Transport{
+			TLSClientConfig: &tls.Config{
+				Certificates: []tls.Certificate{cert},
+			},
+			MaxIdleConns:        10,
+			MaxConnsPerHost:     8,
+			MaxIdleConnsPerHost: 8,
+			IdleConnTimeout:     30 * time.Second,
+			DisableCompression:  true,
+		}
+	}
+	return nil
+}
+
+// Client performs HTTP-based tasks (e.g. upload, download)
+// against a EOS management node (MGM)
+// using the EOS XrdHTTP interface.
+// In this module we wrap eos-related behaviour, e.g. headers or r/w retries
+type Client struct {
+	opt Options
+
+	cl *http.Client
+}
+
+// New creates a new client with the given options.
+func New(opt *Options) *Client {
+	log := logger.New().With().Int("pid", os.Getpid()).Logger()
+	log.Debug().Str("func", "New").Str("Creating new eoshttp client. opt: ", "'"+fmt.Sprintf("%#v", opt)+"' ").Msg("")
+
+	if opt == nil {
+		log.Debug().Str("opt is nil, Error creating http client ", "").Msg("")
+		return nil
+	}
+
+	c := new(Client)
+	c.opt = *opt
+
+	// Let's be successful if the ping was ok. This is an initialization phase
+	// and we enforce the server to be up
+	log.Debug().Str("func", "newhttp").Str("Connecting to ", "'"+opt.BaseURL+"'").Msg("")
+
+	c.cl = &http.Client{
+		Transport: httpTransport}
+
+	c.cl.CheckRedirect = func(req *http.Request, via []*http.Request) error {
+		return http.ErrUseLastResponse
+	}
+
+	if c.cl == nil {
+		log.Debug().Str("Error creating http client ", "").Msg("")
+		return nil
+	}
+
+	return c
+}
+
+// Format a human readable line that describes a response
+func rspdesc(rsp *http.Response) string {
+	desc := "'" + fmt.Sprintf("%d", rsp.StatusCode) + "'" + ": '" + rsp.Status + "'"
+
+	buf := new(bytes.Buffer)
+	r := "<none>"
+	n, e := buf.ReadFrom(rsp.Body)
+
+	if e != nil {
+		r = "Error reading body: '" + e.Error() + "'"
+	} else if n > 0 {
+		r = buf.String()
+	}
+
+	desc += " - '" + r + "'"
+
+	return desc
+}
+
+// If the error is not nil, take that
+// If there is an error coming from EOS, erturn a descriptive error
+func (c *Client) getRespError(rsp *http.Response, err error) error {
+	if err != nil {
+		return err
+	}
+
+	if rsp.StatusCode == 0 {
+		return nil
+	}
+
+	switch rsp.StatusCode {
+	case 0, 200, 201:
+		return nil
+	case 403:
+		return errtypes.PermissionDenied(rspdesc(rsp))
+	case 404:
+		return errtypes.NotFound(rspdesc(rsp))
+	}
+
+	err2 := errtypes.InternalError("Err from EOS: " + rspdesc(rsp))
+	return err2
+}
+
+// From the basepath and the file path... build an url
+func (c *Client) buildFullURL(urlpath, uid, gid string) (string, error) {
+	s := c.opt.BaseURL
+	if len(urlpath) > 0 && urlpath[0] != '/' {
```

Comment:
Can you do something like https://github.com/cs3org/reva/blob/master/pkg/ocm/share/manager/json/json.go#L321-L326?